ykrt/src/location.rs

```diff
@@ -97,88 +29,407 @@ impl State {
-        State(
-            self.state
-                .compare_exchange(current.0, new.0, order, Ordering::Relaxed)
-                .unwrap_or_else(|e| e),
-        )
+    /// Unlocks this `State` with `Release` ordering.
+    pub(crate) fn unlock(&self) {
+        let ls = self.load(Ordering::Relaxed);
+        debug_assert!(ls.is_locked());
+        debug_assert!(!ls.is_counting());
+        if self
+            .state
+            .compare_exchange(
+                ls.with_unparked().x,
+                ls.with_unparked().with_unlock().x,
+                Ordering::Release,
+                Ordering::Relaxed,
+            )
+            .is_ok()
+        {
+            return;
+        }
+
+        // At this point we know another thread has parked itself.
+        let key = unsafe { ls.hot_location() } as *const _ as usize;
+        debug_assert_ne!(key, 0);
+        let callback = |result: UnparkResult| {
+            if result.unparked_threads != 0 && result.be_fair {
+                if !result.have_more_threads {
+                    debug_assert!(ls.is_locked());
+                    self.state.store(ls.with_unparked().x, Ordering::Relaxed);
+                }
+                return TOKEN_HANDOFF;
+            }
+
+            if result.have_more_threads {
+                self.state
+                    .store(ls.with_unlock().with_parked().x, Ordering::Release);
+            } else {
+                self.state
+                    .store(ls.with_unparked().with_unlock().x, Ordering::Release);
+            }
+            TOKEN_NORMAL
+        };
+        unsafe {
+            unpark_one(key, callback);
+        }
     }
 
-    pub(crate) fn store(&self, state: State, order: Ordering) {
-        self.state.store(state.0, order);
+    /// Try obtaining the lock, returning the new `State` if successful.
+    pub(crate) fn try_lock(&self) -> Result<State<I>, ()> {
+        let mut ls = self.load(Ordering::Relaxed);
+        loop {
+            if ls.is_locked() {
+                return Err(());
+            }
+            let new_ls = ls.with_lock();
+            match self.compare_exchange_weak(ls, new_ls, Ordering::Acquire, Ordering::Relaxed) {
+                Ok(_) => return Ok(new_ls),
+                Err(x) => ls = x,
+            }
+        }
     }
 }
 
 impl<I> Drop for Location<I> {
     fn drop(&mut self) {
-        let lp = *self.state.get_mut();
-        if lp & PHASE_TAG == PHASE_TRACING {
-            unsafe { Arc::from_raw((lp & !PHASE_TAG) as *mut u8) };
+        let ls = self.load(Ordering::Relaxed);
+        if !ls.is_counting() {
+            debug_assert!(!ls.is_locked());
+            let hr = unsafe { ls.hot_location() };
+            unsafe {
+                Box::from_raw(hr);
+            }
+        }
+    }
+}
+
+#[cfg(target_pointer_width = "64")]
+pub(crate) const STATE_TAG: usize = 0b111; // All of the other tag data must fit in this.
```

Comment:
Does it make more sense to use a u64? Then it at-least works on 32-bit arches.

TBH, I don't think we care much about those chips, but we can at least allow it to work (albeit slowly).

Of course there are other bits of the system that wouldn't work...

Reply:
> Does it make more sense to use a u64?

No, because these need to encode pointers, so you want to use a native word type -- for CHERI-wide pointers, for example, this would need to be 128 bits.

Reply:
Fair enough.